Add support for update.dpp intrinsic

stefankoncarevic opened the pull request:
"Extended the DPP operation to support the optional 'old' operandand the update.dpp intrinsic. Implemented and ensured 'old' defaults to 'src' when not provided. This allows the use of update.dpp in all cases, effectively behaving like mov.dpp when lowering in backend."

stefankoncarevic committed:
"Extended the DPP operation to support the optional 'old' operand
and the update.dpp intrinsic. Implemented and ensured 'old'
defaults to 'src' when not provided. This allows the use of
update.dpp in all cases, effectively behaving like mov.dpp when
lowering in backend."

stefankoncarevic commented on external/llvm-project/mlir/include/mlir/Dialect/LLVMIR/ROCDLOps.td:
I was guided by this
https://github.com/llvm/llvm-project/blob/c1086532d4d5c0a261457dfb00e79fcb764e3d78/llvm/lib/Target/AMDGPU/VOP1Instructions.td#L1339
to only use update dpp.

krzysz00 replied:
... that's not what that says.

That's two rewrite patterns.

Can you please point me to some specific indication that the mov.dpp intrinsic shouldn't be used?

krzysz00 commented on external/llvm-project/mlir/include/mlir/Dialect/AMDGPU/IR/AMDGPU.td:
```diff
@@ -461,7 +462,7 @@ def AMDGPU_DPPOp : AMDGPU_Op<"dpp", [SameTypeOperands, AllTypesMatch<["result",
   let results = (outs AnyType:$result);
   let assemblyFormat = [{
-    $src $kind (`(` $permArgument^ `)`)? attr-dict `:` type($result)
+    $src (`(` $old^ `:` type($old) `)`)? $kind (`(` $permArgument^ `)`)? attr-dict `:` type($result) 
```
```
src (`,'^ $old)?
```

Also, if you wait for upstream merge to land first, you'll get access to `OptionalTypesMatchWith`, so you don't have to repeat the type.

krzysz00 reviewed (changes requested):
I'd like to see evidence of this alleged deprecation

stefankoncarevic commented:
> I'd like to see evidence of this alleged deprecation

On this link https://github.com/llvm/llvm-project/pull/89233
from the upstream ticket there is a comment from Matt, that mov dpp is deprecated intrinsic, and should be replaced with update.dpp.
The link I posted above in the comment only gives an insight into how mov.dpp can be replaced with update.dpp to behave the same. As for me, I can separate in ROCDLOps.td that we have both mov.dpp and update.dpp so that we can choose which one we want to use.

krzysz00 reviewed (commented):
Per comments upstream, let's do what Matt said and always have an `old` argument and always use `update_dpp`.

However, while you're doing that, can you add some stuff to the documentation on mov.dpp - either in the LLVM documentation or in comments in `IntrinsicsAMDGPU.td` and maybe a deprecation on the builtin in clang? That'll be a separate commit but let's keep it in the same PR here.

stefankoncarevic committed:
"Updated AMDGPU_DPPOp to make 'old' operand mandatory and replaced
usage of mov.dpp with update.dpp due to deprecation of mov.dpp.
Added documentation in the IntrinsicsAMDGPU.td file to reflect
these changes and guide future usage."

krzysz00 reviewed (approved):
Approved!

stefankoncarevic committed:
Resolved deprecation warnings by updating usage of dyn_cast in attribute retrieval.

stefankoncarevic committed:
"Added support for 64-bit operands for update dpp operation."
